_posts/2018-05-07-is-graphql-the-future.markdown

```diff
@@ -0,0 +1,188 @@
+---
+layout: post
+title: Is GraphQL The Future?
+date: 2018-05-07
+author: acjay
+categories: [programming, api, graphql, rest]
+---
+
+I have seen the future, and it looks a lot like GraphQL. Mark my words: in 5 years, newly minted full-stack app developers won’t be debating _RESTfulness_ anymore, because REST API design will be obsolete. By the end of this post, I hope you'll see what I see in the promise of GraphQL as a new approach to client-server interaction.
+
+<!-- more -->
+
+GraphQL is taking the full-stack world by storm. In case you’re not familiar, GraphQL is a language-independent framework for client-server communication. It lets you model the resources and operations provided by a server as a DSL. Clients can use to send scripts written in your DSL to the server to process and respond to as a batch.
+
+That’s...different from how GraphQL’s own page describes it. GraphQL is better known as a query language designed for clients to fetch exactly the data they need. While this is sort of true, I would argue that GraphQL actually fails this test in reality. It’s neither a query language, nor particularly graph-oriented. I argue that it's _not_ a query language because it comes with no native concepts of operators and expressions that build up to queries. _You_ build whatever facilities for specifying and fulfilling queries on your own. Likewise, if your data is a graph, it’s on you to expose that structure. But your requests are, if anything, trees.
+
+I’m not trying to be pedantic. I believe GraphQL succeeds at something subtler and more important than literally being a graph query language. I’m writing this piece because I kept running into difficulties approaching GraphQL from the standpoints of REST, graph theory, or typical query languages. As I read blog posts, StackOverflow Q&As, issues on the GraphQL repo and the GraphQL spec itself, I developed a much nuanced understanding, which I outline below.
+
+For brevity, the following assumes a intermediate familiarity with GraphQL, including its type system, syntax, and server-side implementation. If you don’t have this level of familiarity, I recommend going through any tutorial that requires you to set up a GraphQL server, not just play with the query language (which is how I ended up with a lot of misconceptions). I’m going to start with the basics, but only so I can put my own spin on those concepts, not to really illustrate them with examples.
```

Comment:
I was afraid of that 😄. Defintiely open to suggestions for what resources helped other people, because I built my GraphQL understanding in a very piecemeal fashion.

Reply:
Probably these two: 
- https://www.howtographql.com/basics/0-introduction/
- https://graphql.org/learn/

Reply:
@orta Only problem with those is that they don't seem to walk the user through writing a server :( I think I'll point the user to the GraphQL.js docs: https://graphql.org/graphql-js/